File: Dart/9-DynamicProgramming/Algorithms/3-LongestCommonSubsequence(LCS)/LongestCommonSubsequence(LCS).dart
import 'dart:math';

// Advanced Longest Common Subsequence (LCS) Solutions in Dart
// Implementing five distinct approaches with real-world applications

// 1. Recursive Approach
// This is the most basic approach, directly implementing the recursive definition of LCS.
// It has exponential time complexity due to repeated calculations of the same subproblems.
int lcsRecursive(String s1, String s2, int m, int n) {
  // Base case: If either string is empty, the LCS length is 0.
  if (m == 0 || n == 0) return 0;

  // If the last characters of both strings match, the LCS length increases by 1,
  // and we recurse for the remaining prefixes.
  if (s1[m - 1] == s2[n - 1]) {
    return 1 + lcsRecursive(s1, s2, m - 1, n - 1);
  } else {
    // If the last characters don't match, we take the maximum LCS length from
    // the two possible subproblems:
    // 1. Excluding the last character of s1.
    // 2. Excluding the last character of s2.
    return max(lcsRecursive(s1, s2, m - 1, n), lcsRecursive(s1, s2, m, n - 1));
  }
}

// 2. Memoization Approach (Top-down DP)
// Memoization optimizes the recursive approach by storing the results of subproblems
// in a map (memo) to avoid redundant calculations.
Map<String, int> memo = {};
int lcsMemoization(String s1, String s2, int m, int n) {
  // Base case: If either string is empty, the LCS length is 0.
  if (m == 0 || n == 0) return 0;

  // Create a unique key for the current subproblem.
  String key = "$m-$n";

  // If the result is already in the memo, return it.
  if (memo.containsKey(key)) return memo[key]!;

  // If the last characters of both strings match, the LCS length increases by 1,
  // and we recurse for the remaining prefixes, storing the result in memo.
  if (s1[m - 1] == s2[n - 1]) {
    return memo[key] = 1 + lcsMemoization(s1, s2, m - 1, n - 1);
  }
  // If the last characters don't match, we take the maximum LCS length from
  // the two possible subproblems, storing the result in memo.
  return memo[key] = max(lcsMemoization(s1, s2, m - 1, n), lcsMemoization(s1, s2, m, n - 1));
}

// 3. Iterative Dynamic Programming Approach (Bottom-up DP)
// This approach builds a table (dp) to store the results of subproblems in a bottom-up manner.
// It avoids recursion and typically has better performance than memoization.
int lcsIterative(String s1, String s2) {
  int m = s1.length, n = s2.length;
  // Create a 2D list (dp) to store the LCS lengths for different prefixes of s1 and s2.
  // dp[i][j] represents the LCS length of s1[0...i-1] and s2[0...j-1].
  List<List<int>> dp = List.generate(m + 1, (_) => List.filled(n + 1, 0));

  // Iterate through the characters of s1 and s2.
  for (int i = 1; i <= m; i++) {
    for (int j = 1; j <= n; j++) {
      // If the characters match, the LCS length increases by 1.
      if (s1[i - 1] == s2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        // If the characters don't match, we take the maximum LCS length from
        // the two possible subproblems.
        dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }
  // The result is stored in the bottom-right cell of the dp table.
  return dp[m][n];
}

// 4. Space Optimized DP Approach
// This approach optimizes the iterative DP method by using only two rows (prev and curr)
// instead of the entire dp table. This reduces the space complexity from O(m*n) to O(n).
int lcsSpaceOptimized(String s1, String s2) {
  int m = s1.length, n = s2.length;
  // Create two 1D lists to store the previous and current row values.
  List<int> prev = List.filled(n + 1, 0);
  List<int> curr = List.filled(n + 1, 0);

  // Iterate through the characters of s1 and s2.
  for (int i = 1; i <= m; i++) {
    for (int j = 1; j <= n; j++) {
      // If the characters match, the LCS length increases by 1.
      if (s1[i - 1] == s2[j - 1]) {
        curr[j] = prev[j - 1] + 1;
      } else {
        // If the characters don't match, we take the maximum LCS length from
        // the two possible subproblems.
        curr[j] = max(prev[j], curr[j - 1]);
      }
    }
    // Update the previous row with the current row values.
    prev = List.from(curr);
  }
  // The result is stored in the last row (prev).
  return prev[n];
}

// 5. Real-World Application: DNA Sequence Matching
// This function demonstrates a real-world application of LCS in DNA sequence matching.
// Comparing DNA sequences helps in identifying similarities and differences,
// which is crucial in bioinformatics for tasks like gene alignment and evolutionary analysis.
int lcsDNA(String dna1, String dna2) {
  return lcsIterative(dna1, dna2);
}

void main() {
  String s1 = "AGGTAB";
  String s2 = "GXTXAYB";

  print("Recursive LCS: ${lcsRecursive(s1, s2, s1.length, s2.length)}");
  print("Memoization LCS: ${lcsMemoization(s1, s2, s1.length, s2.length)}");
  print("Iterative DP LCS: ${lcsIterative(s1, s2)}");
  print("Space Optimized LCS: ${lcsSpaceOptimized(s1, s2)}");
  print("DNA Sequence Matching: ${lcsDNA(s1, s2)}");
}

File: C++/9-DynamicProgramming(DP)/Algorithms/3-LongestCommonSubsequence(LCS)/LongestCommonSubsequence(LCS).js
// Longest Common Subsequence (LCS) Problem

// Helper function to find the maximum of two numbers.
function max(a, b) {
    return (a > b) ? a : b;
}

// 1. Recursive Approach for LCS
// This is the most basic approach, directly translating the recursive definition of LCS.
// It has exponential time complexity, making it inefficient for larger inputs.
function lcsRecursive(str1, str2, m, n) {
    if (m === 0 || n === 0) {
        return 0;
    }
    if (str1[m - 1] === str2[n - 1]) {
        return 1 + lcsRecursive(str1, str2, m - 1, n - 1);
    } else {
        return max(lcsRecursive(str1, str2, m - 1, n), lcsRecursive(str1, str2, m, n - 1));
    }
}

// 2. Memoization (Top-Down DP) Approach for LCS
// This approach improves the recursive solution by storing the results of subproblems
// in a table (memo) to avoid redundant computations.
// It has a time complexity of O(m*n), where m and n are the lengths of the strings.
function lcsMemoization(str1, str2, m, n, memo) {
    if (m === 0 || n === 0) {
        return 0;
    }
    if (memo[m][n] !== -1) {
        return memo[m][n];
    }
    if (str1[m - 1] === str2[n - 1]) {
        memo[m][n] = 1 + lcsMemoization(str1, str2, m - 1, n - 1, memo);
        return memo[m][n];
    } else {
        memo[m][n] = max(lcsMemoization(str1, str2, m - 1, n, memo), lcsMemoization(str1, str2, m, n - 1, memo));
        return memo[m][n];
    }
}

// 3. Tabulation (Bottom-Up DP) Approach for LCS
// This approach builds the solution table iteratively, starting from the base cases.
// It also has a time complexity of O(m*n) and is generally more efficient than memoization.
function lcsTabulation(str1, str2, m, n) {
    const dp = Array(m + 1).fill(null).map(() => Array(n + 1).fill(0));

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (str1[i - 1] === str2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }
    return dp[m][n];
}

// 4. Space Optimized Tabulation Approach for LCS
// This approach optimizes the tabulation method by reducing the space complexity
// from O(m*n) to O(min(m, n)).  It only stores the previous row of the DP table.
function lcsSpaceOptimized(str1, str2, m, n) {
    if (m < n) {
        [str1, str2, m, n] = [str2, str1, n, m]; // Ensure m >= n
    }

    let prev = Array(n + 1).fill(0);
    let curr = Array(n + 1).fill(0);

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (str1[i - 1] === str2[j - 1]) {
                curr[j] = 1 + prev[j - 1];
            } else {
                curr[j] = max(prev[j], curr[j - 1]);
            }
        }
        prev = [...curr]; // Copy curr to prev
    }
    return prev[n];
}

// 5.  Finding the LCS String (Tabulation)
function lcsString(str1, str2, m, n) {
    const dp = Array(m + 1).fill(null).map(() => Array(n + 1).fill(0));

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (str1[i - 1] === str2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    let index = dp[m][n];
    let lcs = "";
    let i = m, j = n;
    while (i > 0 && j > 0) {
        if (str1[i - 1] === str2[j - 1]) {
            lcs = str1[i - 1] + lcs;
            i--;
            j--;
            index--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            i--;
        } else {
            j--;
        }
    }
    return lcs;
}

// Example usage:
const str1 = "AGGTAB";
const str2 = "GXTXAYB";
const m = str1.length;
const n = str2.length;

// Initialize memoization table with -1
const memo = Array(m + 1).fill(null).map(() => Array(n + 1).fill(-1));

console.log("String 1: " + str1);
console.log("String 2: " + str2);
console.log("\nResults:");

console.log("1. Recursive LCS Length:", lcsRecursive(str1, str2, m, n));
console.log("2. Memoization LCS Length:", lcsMemoization(str1, str2, m, n, memo));
console.log("3. Tabulation LCS Length:", lcsTabulation(str1, str2, m, n));
console.log("4. Space Optimized LCS Length:", lcsSpaceOptimized(str1, str2, m, n));
console.log("5. LCS String:", lcsString(str1, str2, m, n));
